src/main/java/net/minecraftforge/common/extensions/ForgeRecipeManager.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2018.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.extensions;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.Maps;
+
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.resources.IResourceManagerReloadListener;
+import net.minecraft.util.NonNullList;
+import net.minecraft.world.World;
+import net.minecraftforge.common.crafting.RecipeType;
+
+@SuppressWarnings("deprecation")
+public abstract class ForgeRecipeManager implements IResourceManagerReloadListener
+{
+    protected final Map<RecipeType<? extends IRecipe>, List<? extends IRecipe>> sortedRecipes = Maps.newHashMap();
+
+    @SuppressWarnings("unchecked")
+    public <T extends IRecipe> List<T> getRecipes(RecipeType<T> type)
+    {
+        return (List<T>) this.sortedRecipes.computeIfAbsent(type, t -> new ArrayList<>());
+    }
+
+    public Collection<IRecipe> getRecipes(Collection<RecipeType<?>> types)
+    {
+        Set<IRecipe> recipes = new HashSet<>();
+        for(RecipeType<?> t : types) recipes.addAll(getRecipes(t));
+        return recipes;
+    }
+
+    public ItemStack getResult(IInventory input, World world, RecipeType<?> type)
+    {
+        for(IRecipe irecipe : getRecipes(type))
```

Comment:
Perfect opertunity for one line stream functions.

Reply:
I'm avoiding streams for performance concerns.  While this method specifically isn't used much in vanilla, it's better to be safe if someone actually starts using it.